Sources/KDTree.swift

```diff
@@ -62,27 +61,104 @@ extension KDTree {
             //Needed for 'contains' and 'removing' method.
-            while median >= 1 && abs(sortedValues[median-1].kdDimension(currentSplittingDimension) - medianValue) < Double.ulpOfOne {
+            while median >= 1 && median > selected.startIndex && abs(selected[median-1].kdDimension(currentSplittingDimension) - medianValue) < Double.ulpOfOne {
                 median -= 1
             }
             
-            let leftTree = KDTree(values: sortedValues[0..<median], depth: depth+1)
-            let rightTree = KDTree(values: sortedValues[median+1..<sortedValues.count], depth: depth+1)
+            let leftTree = KDTree(values: selected[selected.startIndex..<median], depth: depth+1)
+            let rightTree = KDTree(values: selected[median+1..<selected.endIndex], depth: depth+1)
             
-            self = .node(left: leftTree, value: sortedValues[median],
+            self = .node(left: leftTree, value: selected[median],
                          dimension: currentSplittingDimension, right: rightTree)
         }
     }
     
+    /*
+     Lomuto's partitioning algorithm.
+     This is conceptually simpler than Hoare's original scheme but less efficient.
+     The return value is the index of the pivot element in the new array. The left
+     partition is [low...p-1]; the right partition is [p+1...high], where p is the
+     return value.
+     The left partition includes all values smaller than the pivot, so
+     if the pivot value occurs more than once, its duplicates will be found in the
+     right partition.
+     */
+    private static func partitionLomuto(_ a: inout ArraySlice<Element>, kdDimension: Int) -> Int {
+        // We always use the highest item as the pivot.
+        let pivot = a.last!
+        
+        // This loop partitions the array into four (possibly empty) regions:
+        //   [a.startIndex  ...   i] contains all values < pivot,
+        //   [i+1  ...          j-1] contains all values >= pivot,
+        //   [j    ..< a.endIndex-1] are values we haven't looked at yet,
+        //   [a.endIndex           ] is the pivot value.
+        var i = a.startIndex
+        
+        for j in a.startIndex..<a.endIndex-1 {
+            if a[j].kdDimension(kdDimension) < pivot.kdDimension(kdDimension) {
+                (a[i], a[j]) = (a[j], a[i])
+                i += 1
+            }
+        }
+        
+        // Swap the pivot element with the first element that is greater than
+        // the pivot. Now the pivot sits between the < and >= regions and the
+        // array is properly partitioned.
+        (a[i], a[a.endIndex-1]) = (a[a.endIndex-1], a[i])
+        return i
+    }
+    
+    /*
+     takes an array and an index n and returns an array with the nth value being the nth largest
+     */
+    private static func quickSelect(n: Int, arr: [Element], kdDimension: Int) -> [Element] {
+        if arr.count == 1 {
+            return [arr.first!]
```

Comment:
I think this might be a good place for a `guard` statement (Most of the code I've worked o and written in the last two years only uses `guard` for early exists).
I.e. a one-liner is very common

```
guard arr.count > 1 else { return arr }
```